### ISSUE TITLE
 if stride!=1 or bottom.shape[1] != num_filter*widen_factor: ('Top' object has no attribute 'shape')

### ISSUE THREAD

**Issue by congmonkey** [bob48523/modified_wrn#3]
# -*- coding: utf-8 -*-
"""
Created on Sat Aug 19 15:19:32 2017

@author: chenjiaxu
"""

from __future__ import print_function

import math
from caffe import layers as L, params as P, to_proto
from caffe.proto import caffe_pb2
import caffe
#model = L.Eltwise(model, conv1,operation = 'SUM')
def bn_relu_conv(bottom, ks, nout, stride, pad, dropout):
    batch_norm = L.BatchNorm(bottom, in_place=False, param=[dict(lr_mult=0, decay_mult=0), dict(lr_mult=0, decay_mult=0), dict(lr_mult=0, decay_mult=0)])
    scale = L.Scale(batch_norm, bias_term=True, in_place=True, filler=dict(value=1), bias_filler=dict(value=0))
    relu = L.ReLU(scale, in_place=True)
    conv = L.Convolution(relu, kernel_size=ks, stride=stride, 
                    num_output=nout, pad=pad, bias_term=False, weight_filler=dict(type='msra'), bias_filler=dict(type='constant'))
    if dropout>0:
        conv = L.Dropout(conv, dropout_ratio=dropout)
    return conv

def preactbottleneck(bottom, ks, nout, stride, pad, groups=1):
    batch_norm = L.BatchNorm(bottom, in_place=False, param=[dict(lr_mult=0, decay_mult=0), dict(lr_mult=0, decay_mult=0), dict(lr_mult=0, decay_mult=0)])
    scale = L.Scale(batch_norm, bias_term=True, in_place=True, filler=dict(value=1), bias_filler=dict(value=0))
    relu = L.ReLU(scale, in_place=True)
    conv = L.Convolution(relu, kernel_size=ks, stride=stride, group=groups,
                    num_output=nout, pad=pad, bias_term=False, weight_filler=dict(type='msra'), bias_filler=dict(type='constant'))
   
    return conv

def shortcut(bottom, nout, stride):
    conv = L.Convolution(bottom, kernel_size=1, stride=stride, 
                    num_output=nout, pad=0, bias_term=False, weight_filler=dict(type='msra'), bias_filler=dict(type='constant'))
   
    return conv
  
def add_layer(bottom, num_filter, stride):
    widen_factor = 4
    
    if stride!=1 or bottom.shape[1] != num_filter*widen_factor:
        x = shortcut(bottom, nout=num_filter*widen_factor, stride=stride)
    else:
        x = bottom
    
    conv = preactbottleneck(bottom, ks=1, nout=num_filter, stride=1, pad=0, groups=1)
    conv = preactbottleneck(conv, ks=3, nout=num_filter, stride=stride, pad=1, groups=16)
    conv = preactbottleneck(conv, ks=1, nout=num_filter*widen_factor, stride=1, pad=0, groups=16)
    
    out = L.Eltwise(x, conv)
    return out

def transition(bottom, num_filter, dropout):
    conv = bn_relu_conv(bottom, ks=1, nout=num_filter, stride=1, pad=0, dropout=dropout)
    pooling = L.Pooling(conv, pool=P.Pooling.AVE, kernel_size=2, stride=2)
    return pooling

#change the line below to experiment with different setting
#depth -- must be 3n+4
#first_output -- channels before entering the first dense block, twice the growth_rate for DenseNet-BC
#growth_rate -- growth rate
#dropout -- set to 0 to disable dropout, non-zero number to set dropout rate
def resnetx(data_file, mode='train', batch_size=64, depth=[3,4,6,3], width=[32,64,128,256]):
    data, label = L.Data(source=data_file, backend=P.Data.LMDB, batch_size=batch_size, ntop=2, 
              transform_param=dict(mean_file="/data1/zuotongchun/cifar100_mean/cifar100_train_all_mean.binaryproto"))

    nchannels = 32
    model = L.Convolution(data, kernel_size=3, stride=1, num_output=nchannels,
                        pad=1, bias_term=False, weight_filler=dict(type='msra'), bias_filler=dict(type='constant'))
    model = L.BatchNorm(model, in_place=False, param=[dict(lr_mult=0, decay_mult=0), dict(lr_mult=0, decay_mult=0), dict(lr_mult=0, decay_mult=0)])
    # 1 x 1 convolution produce 4k feature-maps
    # nchannels_B = 4 * growth_rate
    #N1 = 6
    #N2 = 12
    #N3 = 32
    #N4 = 32
    strides = [1] + [1]*(depth[0]-1)
    for stride in strides:        
        model = add_layer(model, width[0], stride)
        
    strides = [2] + [1]*(depth[1]-1)
    for stride in strides:        
        model = add_layer(model, width[1], stride)

    strides = [2] + [1]*(depth[2]-1)
    for stride in strides:        
        model = add_layer(model, width[2], stride)
        
    strides = [2] + [1]*(depth[3]-1)
    for stride in strides:        
        model = add_layer(model, width[3], stride)
        
    model = L.Pooling(model, pool=P.Pooling.AVE, global_pooling=True)
    model = L.InnerProduct(model, num_output=10, bias_term=True, weight_filler=dict(type='xavier'), bias_filler=dict(type='constant'))
    loss = L.SoftmaxWithLoss(model, label)
    accuracy = L.Accuracy(model, label)
    return to_proto(loss, accuracy)

def make_net():

    with open('train-resnetx-BC-50.prototxt', 'w') as f:
        # change the path to your data. If it's not lmdb format, also change first line of densenet() function
        print(str(resnetx('/data1/zuotongchun/cifar10_train_all_lmdb', batch_size=64)), file=f)

    with open('test-resnetx-BC-50.prototxt', 'w') as f:
        print(str(resnetx('/data1/zuotongchun/cifar10_test_lmdb', batch_size=50)), file=f)

def make_solver():
    s = caffe_pb2.SolverParameter()
    s.random_seed = 0xCAFFE

    s.train_net = 'train-resnetx-BC-50.prototxt'
    s.test_net.append('train-resnetx-BC-50.prototxt')
    s.test_interval = 200
    s.test_iter.append(200)

    s.max_iter = 230000
    s.type = 'Nesterov'
    s.display = 100

    s.base_lr = 0.1
    s.momentum = 0.9
    s.weight_decay = 1e-1

    s.lr_policy='multistep'
    s.gamma = 0.1
    s.stepvalue.append(int(0.5 * s.max_iter))
    s.stepvalue.append(int(0.75 * s.max_iter))
    s.solver_mode = caffe_pb2.SolverParameter.GPU

    # add by sdh
    s.device_id = 6 # 2 or 3
    s.snapshot = 1000
    s.snapshot_prefix = 'snapshot/densenet-BC-169'

    solver_path = 'solver.prototxt'
    with open(solver_path, 'w') as f:
        f.write(str(s))
        f.write('\n')

if __name__ == '__main__':

    make_net()
    make_solver()
